proto_proxy.c

```diff
@@ -0,0 +1,2310 @@
+            // set the head here. when we break the head will be correct.
+            STAILQ_REMOVE_HEAD(&be->io_head, io_next);
+            be->depth--;
+            if (STAILQ_EMPTY(&be->io_head)) {
+                // TODO: suspicious of this code. audit harder?
+                stop = true;
+            }
+
+            // have to do the q->count-- and == 0 and redispatch_conn()
+            // stuff here. The moment we call that write we
+            // don't own *p anymore.
+            // FIXME: are there any other spots where p->q or p's p->next
+            // stack are examined from what would be multiple servers at once?
+            p->q->count--;
+            if (p->q->count == 0) {
+                redispatch_conn(p->c);
+            }
+
+            // mcmc_buffer_consume() - if leftover, keep processing
+            // IO's.
+            // if no more data in buffer, need to re-set stack head and re-set
+            // event.
+            remain = 0;
+            // TODO: do we need to yield every N reads?
+            newbuf = mcmc_buffer_consume(be->client, &remain);
+            if (remain != 0) {
+                // data trailing in the buffer, for a different request.
+                memmove(be->rbuf, newbuf, remain);
+                //P_DEBUG("read buffer remaining: %d\n", remain);
+            } else {
+                // TODO: STAILQ_EMPTY() sets stop = true above.
+                // remove this and re-attempt a read?
+                stop = true;
+            }
+
+            be->state = mcp_backend_read;
+
+            break;
+        default:
+            // TODO: at some point (after v1?) this should attempt to recover,
+            // though we should only get here from memory corruption and
+            // bailing may be the right thing to do.
+            fprintf(stderr, "%s: invalid backend state: %d\n", __func__, be->state);
+            assert(false);
+    } // switch
+    } // while
+
+    return flags;
+}
+
+// TODO: add a second argument for assigning a specific error to all pending
+// IO's (ie; timeout).
+// The backend has gotten into a bad state (timed out, protocol desync, or
+// some other supposedly unrecoverable error: purge the queue and
+// cycle the socket.
+// Note that some types of errors may not require flushing the queue and
+// should be fixed as they're figured out.
+// _must_ be called from within the event thread.
+static int _reset_bad_backend(mcp_backend_t *be) {
+    io_pending_proxy_t *io = NULL;
+    STAILQ_FOREACH(io, &be->io_head, io_next) {
+        // TODO: Unsure if this is the best way of surfacing errors to lua,
+        // but will do for V1.
+        io->client_resp->status = MCMC_ERR;
+        io->q->count--;
+        if (io->q->count == 0) {
+            redispatch_conn(io->c);
+        }
+    }
+
+    STAILQ_INIT(&be->io_head);
+
+    mcmc_disconnect(be->client);
+    int status = mcmc_connect(be->client, be->ip, be->port, MCMC_OPTION_NONBLOCK);
```

Comment:
Planned but not for my "v1" stuff here. I need the structure required for validating servers pre-usage (ie `version` check) and then piggyback on that since it's async. I also need to ensure users can avoid relying on DNS so raw IP's will still work (not everyone has a reliable DNS service sadly).